aws-rds-cfn-common/src/main/java/software/amazon/rds/common/logging/ProxyClientLogger.java

```diff
@@ -0,0 +1,92 @@
+package software.amazon.rds.common.logging;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import software.amazon.awssdk.awscore.AwsRequest;
+import software.amazon.awssdk.awscore.AwsResponse;
+import software.amazon.awssdk.core.ResponseBytes;
+import software.amazon.awssdk.core.ResponseInputStream;
+import software.amazon.awssdk.core.pagination.sync.SdkIterable;
+import software.amazon.cloudformation.proxy.ProxyClient;
+
+public class ProxyClientLogger {
+
+    public static <ClientT> ProxyClient<ClientT> newProxy(final RequestLogger requestLogger,
+                                                          final ProxyClient<ClientT> proxyClient) {
+        return new ProxyClient<ClientT>() {
+            @Override
+            public <RequestT extends AwsRequest, ResponseT extends AwsResponse>
+            ResponseT
+            injectCredentialsAndInvokeV2(RequestT request, Function<RequestT, ResponseT> requestFunction) {
+                ResponseT result = getResponseT(requestLogger, request, requestFunction,
+                        (req, reqFunction) -> proxyClient.injectCredentialsAndInvokeV2(req, reqFunction));
+                requestLogger.log(result.getClass().getSimpleName(), result);
+                return result;
+            }
```

Comment:
Looks pretty heavy. Any chance we can solve it using an annotation?

Reply:
Done